pkg/astnormalization/datatype_deduplication.go

```diff
@@ -0,0 +1,72 @@
+package astnormalization
+
+import (
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astvisitor"
+)
+
+func dataTypeDeduplication(walker *astvisitor.Walker) {
+	visitor := dataTypeDeduplicationVisitor{
+		Walker: walker,
+	}
+	walker.RegisterScalarTypeDefinitionVisitor(&visitor)
+	walker.RegisterEnterDocumentVisitor(&visitor)
+	walker.RegisterLeaveDocumentVisitor(&visitor)
+	walker.RegisterEnterScalarTypeDefinitionVisitor(&visitor)
+}
+
+type dataTypeDeduplicationVisitor struct {
+	*astvisitor.Walker
+	operation    *ast.Document
+	definition   *ast.Document
+	typesSeen    map[string]bool
+	refsToRemove []int
+	lastRef      int
+}
+
+func (d *dataTypeDeduplicationVisitor) EnterDocument(operation, definition *ast.Document) {
+	d.operation, d.definition = operation, definition
+	d.typesSeen = make(map[string]bool)
+	d.refsToRemove = make([]int, 0)
+	d.lastRef = -1
+}
+
+func (d *dataTypeDeduplicationVisitor) LeaveDocument(operation, definition *ast.Document) {
+	if len(d.refsToRemove) < 1 {
+		return
+	}
+	d.operation, d.definition = operation, definition
+	newRootNodes := make([]ast.Node, 0)
+MainLoop:
+	for _, node := range d.operation.RootNodes {
+		if node.Kind == ast.NodeKindScalarTypeDefinition {
+			for i, ref := range d.refsToRemove {
+				if node.Ref == ref {
+					lastIndex := len(d.refsToRemove) - 1
+					d.refsToRemove[i] = d.refsToRemove[lastIndex]
+					d.refsToRemove = d.refsToRemove[:lastIndex]
+					continue MainLoop
```

Comment:
I'm not sure if need this continue statement with GOTO here. Why to go though RootNodes again and again? 
To me it feels more logical to have `break` on this line, so it will move on to the next root node.

Reply:
I see the point now, so you need to skip the upper loop, but even in this case `MainLoop:` sits in a wrong place I think. What you can do is change it to a simpler logic like:

```
for _, node := range d.operation.RootNodes {
                skip := false
 		if node.Kind == ast.NodeKindScalarTypeDefinition {
 			for i, ref := range d.refsToRemove {
 				if node.Ref == ref {
 					lastIndex := len(d.refsToRemove) - 1
 					d.refsToRemove[i] = d.refsToRemove[lastIndex]
 					d.refsToRemove = d.refsToRemove[:lastIndex]
                                        skip = true
 					break
                        }
                  }

                  if !skip {
                         newRootNodes = append(newRootNodes, node)
                  }
}
```